File: DsaWithCpp/Tree/BinaryTree/SerializeAndDeserialize.cpp
#include <iostream>
#include <vector>
#include "Node.cpp"
// #include "TreeBuilder.cpp"
#include "TreeTraversals.cpp"
using namespace std;

static void serialize(Node *root, vector<int> &arr)
{
    if (root == NULL)
    {
        arr.push_back(-1);
        return;
    }
    arr.push_back(root->data);
    serialize(root->left, arr);
    serialize(root->right, arr);
}

static int index = 0;

static Node *deSerialize(vector<int> &arr)
{
    if (index == arr.size())
        return NULL;
    int data = arr.at(index++);
    if (data == -1)
        return NULL;
    Node *root = new Node(data);
    root->left = deSerialize(arr);
    root->right = deSerialize(arr);
    return root;
}

int main()
{
    Traversals ob;
    vector<int> arr;
    int nodes[] = {1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1};
    Node *root = buildTree(nodes);
    serialize(root, arr);
    for (int i = 0; i < arr.size(); i++)
        cout << arr[i] << " ";
    cout << endl;
    Node *newRoot = deSerialize(arr);
    ob.preOrder(newRoot);
    cout << endl;
}

File: Python/tree/binary_tree/serialize_ and_deserialize.py
from node import Node
from tree_builder import build_tree
from tree_traversals import pre_order


def serialize(root: Node, arr: list):
    if root is None:
        arr.append(-1)
        return
    arr.append(root.data)
    serialize(root.left, arr)
    serialize(root.right, arr)


index = 0


def deserialize(arr: list) -> Node:
    global index
    if index == len(arr):
        return None
    data = arr[index]
    index += 1
    if data == -1:
        return None
    root = Node(data)
    root.left = deserialize(arr)
    root.right = deserialize(arr)
    return root


def main():
    arr = []
    nodes = [1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1]
    root = build_tree(nodes)
    serialize(root, arr)
    for i in arr:
        print(i, end=" ")
    print()
    new_root = deserialize(arr)
    pre_order(new_root)
    print()


if __name__ == "__main__":
    main()
